experimental/Schemes/BlowupMorphism.jl

```diff
@@ -96,108 +96,236 @@ function strict_transform(p::BlowupMorphism, inc::CoveredClosedEmbedding)
-#  p_cov_simp = compose(phi, p_cov_ref)
-  p_cov_simp = p_cov_ref
-  CY_simp = CY_ref
-  E = exceptional_divisor(p)
-  multipl = -1
-  for U in patches(CY_simp)
-    p_res = p_cov_simp[U]
-    V = codomain(p_res)
-    length(C(V))==1 || error("ideal for divisor is not principal")
-    h = C(V)[1]
-    pbh = pullback(p_res)(h)
-    if isunit(pbh) 
+  pr = projection(p)
+  pr_refined = restrict(pr,CX)::CoveringMorphism
+  CY = domain(pr_refined)
+
+  ## do the transform on the charts
+  multEInC = -1
+  for U in patches(CY)
+    V = codomain(p_refined[U])        # affine patch on X
+
+    ## determine single generator of Cartier divisor C on V
+    length(C(V)) == 1 || error("ideal for divisor is not principal")
+                       # sanity check -- we are on a trivializing covering after all!
+    h_orig = C(V)[1]
+    h_total = pullback(p_refined)(h_orig)
+    if isunit(h_total)
       ID[U] = one(OO(U))
       continue
     end
-    k = 0
-    e = first(E(U))
+
+    ## determine single generator of Cartier divisor E on U
+    length(E(U)) == 1 || error("exceptional divisor is not principal")
+                       # sanity check -- default covering of Y is already trivializing for E!
+    e = E(U)[1]
     if isunit(e)
-      ID[U] = pbh
+      ID[U] = h_total
       continue
     end
-    # Warning! Successive division by e only gives the correct result in this 
-    # particular situation! The equation for e must be irreducible, etc.
-    success, b = divides(pbh, e)
-    while success
-      k = k + 1
-      pbh = b
-      success, b = divides(b, e)
-    end
-    ID[U] = pbh
-    if multipl != -1
-      k == multipl || error("multiplicities differ in different charts")
+
+    ## now it is just division of ring elements
+    h_strict, b = divide_by_max_power(h_total,e,multEInC)
+    if multEinC == -1
+      multEInC = b
     end
-    multipl = k
+
+    ## fill in data of C_strict
+    ID[U] = h_strict
   end
 
-  multipl = (multipl == -1 ? 0 : multipl)
+  ## we are good to go now
+  C_strict = EffectiveCartierDivisor(Y, ID, check=false)
+  return C_strict, multEInC
+end
 
-  C_trans = EffectiveCartierDivisor(Y, ID, check=false) # TODO: Set to false
-  return C_trans # TODO: Do we want to also return the exceptional component multipl*E ?
+#@Markdown.doc """
+#  divide_by_max_power(f::T,g::T,k::Int) where T <: RingElem
+#
+#For two ring elements ``f`` and ``g`` of the same ring and an integer return a pair of a ring element ``h`` and and integer ``m`` such that ``f = h g^m`` and ``g`` does not divide ``h ``.
+#If ``k==-1``, ``m`` is determined; if ``k > 0`` throw an error if ``k != m``.
+#"""
+function divide_by_max_power(f::T,g::T,k::Int,check::Bool=true) where T <: RingElem
+  (k > -2 && k < deg(f)+1) || error("invalid value for k")
+
+  if k == -1
+
+    ftemp = f
+    m = 0
+    ## divide step by step
+    while true
+      success, b = divides(ftemp,g)
+      success || break
+      m = m+1
+      ftemp = b
+    end
```

Comment:
This is doing things iteratively. Alternatively, consider using `minimal_power_such_that` or `maximal_power_such_that` (which probably doesn't exist yet, but can easily be ported from the other implementation).

Reply:
obsolete due to algorithmic change